schema_salad/tests/test_codegen_errors.py

```diff
@@ -0,0 +1,220 @@
+    assert path
+    with pytest.raises(ValidationException, match=match):
+        load_document_by_uri(path)
+
+
+def test_error_message2() -> None:
+    t = "test_schema/test2.cwl"
+    match = r"""^.*test2\.cwl:2:1:\s+Field\s+`class`\s+contains\s+undefined\s+reference\s+to
+\s+`file://.+/schema_salad/tests/test_schema/xWorkflow`$"""
+    path = get_data("tests/" + t)
+    assert path
+    with pytest.raises(ValidationException, match=match):
+        load_document_by_uri(path)
+
+
+def test_error_message4() -> None:
+    t = "test_schema/test4.cwl"
+    match = r"""^.*test4.cwl:2:1:\s+Object\s+`.*test4.cwl`\s+is\s+not\s+valid\s+because
+\s+tried\s+`Workflow`\s+but
+.*test4\.cwl:6:1:\s+the\s+`outputs`\s+field\s+is\s+not\s+valid\s+because:
+.*test4\.cwl:7:3:\s+checking\s+object\s+`.*test4\.cwl#bar`
+\s+tried\s+`WorkflowOutputParameter`\s+but
+\s+the\s+`type`\s+field\s+is\s+not\s+valid\s+because:
+\s+Expected\s+one\s+of\s+\(list|dict|str,list|dict|str,list|dict|str\)\s+was
+\s+<class\s+'int'>"""
+    path = get_data("tests/" + t)
+    assert path
+    with pytest.raises(ValidationException, match=match):
+        load_document_by_uri(path)
+
+
+def test_error_message5() -> None:
+    t = "test_schema/test5.cwl"
+    match = r"""^.*test5\.cwl:2:1:\s+Object\s+`.*test5\.cwl`\s+is\s+not\s+valid\s+because
+\s+tried\s+`Workflow`\s+but
+.+test5\.cwl:8:1:\s+the\s+`steps`\s+field\s+is\s+not\s+valid\s+because:
+\s+tried\s+`array<WorkflowStep>`\s+but
+\s+-\s+tried\s+`array<WorkflowStep>`\s+but
+\s+Expected\s+a\s+list,\s+was\s+<class\s+'int'>
+\s+-\s+tried\s+`WorkflowStep`\s+but
+\s+Expected\s+a\s+dict,\s+was\s+<class\s+'int'>"""
+    path = get_data("tests/" + t)
+    assert path
+    with pytest.raises(ValidationException, match=match):
+        load_document_by_uri(path)
+
+
+def test_error_message6() -> None:
+    t = "test_schema/test6.cwl"
+    match = r"""-\s+tried\s+`CommandLineTool`\s+but
+\s+Missing\s+'class'\s+field
++-\s+tried\s+`ExpressionTool`\s+but
+\s+Missing\s+'class'\s+field
++-\s+tried\s+`Workflow`\s+but
+\s+Missing\s+'class'\s+field"""
+    path = get_data("tests/" + t)
+    assert path
+    with pytest.raises(ValidationException, match=match):
+        load_document_by_uri(path)
+
+
+def test_error_message7() -> None:
+    t = "test_schema/test7.cwl"
+    match = r"""^.*test7\.cwl:2:1:\s+Object\s+`.*test7\.cwl`\s+is\s+not\s+valid\s+because
+\s+tried\s+`Workflow`\s+but
+.*test7\.cwl:8:1:\s+the\s+`steps`\s+field\s+is\s+not\s+valid\s+because:
+\s+tried\s+`array<WorkflowStep>`\s+but
+.*test7\.cwl:9:3:\s+checking\s+object\s+`.*test7.cwl#step1`
+\s+tried\s+`WorkflowStep`\s+but
+\s+\*\s+missing\s+required\s+field\s+`run`
+.*test7\.cwl:10:5:\s+\*\s+invalid\s+field\s+`scatter_method`,\s+expected\s+one of:.*\s+.*\s+.*\s+.*\s+.*\s+.*\s+.*\s+.*\s+.*\s+.*$"""
+    path = get_data("tests/" + t)
+    assert path
+    with pytest.raises(ValidationException, match=match):
+        load_document_by_uri(path)
+
+
+def test_error_message8() -> None:
```

Comment:
Has the redundant `tried 'X' but Y` that I mentioned earlier.

```
tests/test_schema/test8.cwl:2:1:   Object `test8.cwl` is not valid because
                                    tried `Workflow` but
tests/test_schema/test8.cwl:8:1:       the `steps` field is not valid because:
                                        tried `array<WorkflowStep>` but
tests/test_schema/test8.cwl:9:3:           checking object `test8.cwl#step1`
                                            tried `WorkflowStep` but
                                              * missing required field `run`
tests/test_schema/test8.cwl:10:5:             the `scatterMethod` field is not valid because:
                                                contains undefined reference to
                                                `file:///home/peter/work/schema_salad/schema_salad/tests/test_schema/abc`
```

The original validator:

```
test_schema/test8.cwl:8:1:  checking field 'steps'
test_schema/test8.cwl:9:3:    checking object 'test_schema/test8.cwl#step1'
test_schema/test8.cwl:10:5:     Field 'scatterMethod' contains undefined reference to
                                'file:///home/peter/work/schema_salad/schema_salad/tests/test_schema/abc'
```

* the codegen one is actually slightly better, because it notices the missing `run` field
* However, for some reason it has a bullet point in front of `missing required field` but not `the 'scatterMethod` field is not valid`.

So, it should look like this:

```
tests/test_schema/test8.cwl:2:1:   Object `test8.cwl` is not valid because
tests/test_schema/test8.cwl:8:1:       the `steps` field is not valid because:
tests/test_schema/test8.cwl:9:3:           checking object `test8.cwl#step1`
                                            tried `WorkflowStep` but
                                              * missing required field `run`
tests/test_schema/test8.cwl:10:5:             * the `scatterMethod` field is not valid because:
                                                contains undefined reference to
                                                `file:///home/peter/work/schema_salad/schema_salad/tests/test_schema/abc`
```